### PR TITLE
Add new replace_domain action to allow adding domain(s) exclusively to a specific list

### PR DESCRIPTION
**Pull request by DL6ER**
**By submitting this pull request, I confirm the following:** 

- [X] I have read and understood the [contributors guide](https://github.com/pi-hole/pi-hole/blob/master/CONTRIBUTING.md), as well as this entire template.
- [X] I have made only one major change in my proposed changes.
- [X] I have commented my proposed changes within the code.
- [X] I have tested my proposed changes, and have included unit tests where possible.
- [X] I am willing to help maintain this change if there are issues with it later.
- [X] I give this submission freely and claim no ownership.
- [X] It is compatible with the [EUPL 1.2 license](https://opensource.org/licenses/EUPL-1.1)
- [X] I have squashed any insignificant commits. ([`git rebase`](http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html))

---
**What does this PR aim to accomplish?:**

Fix different actions of `pihole -w/b` and the White-/Blacklist buttons on the Query Log (web interface).

**How does this PR accomplish the above?:**

Add new `replace_domain` action in `groups.php` to allow adding domain(s) exclusively to a specific list. In this mode, any occurrences of said domain are first removed from the list before adding the new ones.

**What documentation changes (if any) are needed to support this PR?:**

None

### CODE REVIEW

**Review by PromoFaux — DISMISSED** [review]
Does what it says on the tin

**Review by yubiuser — DISMISSED** [review]

**Comment by DL6ER** [comment]
I'm going to convert this into a draft for now as the underlying issue is likely more severe and we may want to chose another approach, even.

**Comment by DL6ER** [comment]
This is supposed to prevent users from being able to change the domains if there are custom group settings in place. Also, this ensures that the `DELETE` trigger is not fired on updated (already existing) domains by switching from `REPLACE` to `INSERT OR IGNORE + UPDATE`

**Review by PromoFaux — CHANGES_REQUESTED** [review]
image

Only thing to note in the above screenshot is a missing space between the domain and the word "is". Also it is automatically dismissed quite quickly. I had to screenshot to read it fully, can we tweak that?

File: scripts/pi-hole/php/groups.php [hunk 1]
```diff
@@ -497,18 +497,41 @@ function JSON_error($message = null)
...
+        $update_stmt = $db->prepare('UPDATE domainlist SET comment = :comment WHERE domain = :domain AND type = :type');
+        if (!$update_stmt) {
             throw new Exception('While preparing statement: ' . $db->lastErrorMsg());
         }
 
+        $check_stmt = null;
+        $delete_stmt = null;
+        if($_POST['action'] == 'replace_domain') {
+            // Check statement will reveal any group associations for a given (domain,type) which do NOT belong to the default group
+            $check_stmt = $db->prepare('SELECT EXISTS(SELECT * FROM domainlist_by_group WHERE domainlist_id = (SELECT id FROM domainlist WHERE domain = :domain) AND group_id != 0)');
```
Comment (PromoFaux) [inline scripts/pi-hole/php/groups.php]: Now that I am awake again, just looking over the `SQL`. Personally not a fan of nested subqueries, can we do a `JOIN` instead please?

```suggestion
            $check_stmt = $db->prepare('SELECT EXISTS(SELECT domain FROM domainlist_by_group dlbg JOIN domainlist dl on dlbg.domainlist_id = dl.id WHERE dl.domain = :domain AND dlbg.group_id != 0)');
```

**Review by PromoFaux — CHANGES_REQUESTED** [review]

File: scripts/pi-hole/php/groups.php (hunk 1 above)
Comment (PromoFaux) [reply scripts/pi-hole/php/groups.php]: (this should also be more performant)

Comment (DL6ER) [reply scripts/pi-hole/php/groups.php]: Who am I to speak up against the godfather of SQL?

**Comment by DL6ER** [comment]
Force-pushed to get the correct `SignedOff` into the online committed code suggestion by @PromoFaux

**Comment by yubiuser** [comment]
Something is not right yet.
Added a domain `test.com` to the blacklist (default group). Requested the domain, clicked on whitelist in query log. Now I have `test.com` as black and whitelist entry in default group. Shouldn't it replace the domain if it is in default group?

```
nanopi@nanopi:~$ sqlite3 /etc/pihole/gravity.db -header -column "Select * from domainlist where domain='test.com';"
id          type        domain      enabled     date_added  date_modified  comment             
----------  ----------  ----------  ----------  ----------  -------------  --------------------
131         0           test.com    1           1604953867  1604953867     Added from Query Log
130         1           test.com    1           1604953571  1604953571                         
nanopi@nanopi:~$ sqlite3 /etc/pihole/gravity.db -header -column "Select * from domainlist_by_group;"
domainlist_id  group_id  
-------------  ----------
130            0         
131            0   

```

Will do more tests tomorrow

**Comment by DL6ER** [comment]
@yubiuser I just did exactly what you did, however, no issues for me?

1. Checking there is no such domain in my database
    
    ```sql
    SELECT * FROM domainlist JOIN domainlist_by_group ON domainlist_by_group.domainlist_id = domainlist.id WHERE domain = 'test.com'
    ```
    ---> No result
    
2. Adding the domain
    
    ```bash
    $ pihole -b test.com                                                                                                             
      [i] Adding test.com to the blacklist...                                                                                                                                
      [✓] Reloading DNS lists
    ```
    
3. Checking again the database
    
    ```sql
    SELECT * FROM domainlist JOIN domainlist_by_group ON domainlist_by_group.domainlist_id = domainlist.id WHERE domain = 'test.com'
    ```
    ```plain
    id          type        domain      enabled     date_added  date_modified  comment     domainlist_id  group_id
    ----------  ----------  ----------  ----------  ----------  -------------  ----------  -------------  ----------
    243         1           test.com    1           1604955416  1604955416                 243            0
    ```
    
    ---> Only type 1 here
    
4. Creating a query to this domain (and verifying it is blocked)
    
    ```bash
    $ dig +short test.com
    0.0.0.0
    ```
    
5. Query Log: clicked on 'Whitelist' next to this query
    
6. Checking again the database
    ```bash
    $ sqlite3 /etc/pihole/gravity.db -header -column "SELECT * FROM domainlist JOIN domainlist_by_group ON domainlist_by_group.domainlist_id = domainlist.id WHERE domain = 'test.com'"
    id          type        domain      enabled     date_added  date_modified  comment               domainlist_id  group_id
    ----------  ----------  ----------  ----------  ----------  -------------  --------------------  -------------  ----------
    244         0           test.com    1           1604955559  1604955559     Added from Query Log  244            0
    ```
    
    ---> Only type 0 exists after moving to the Whitelist on the Query Log page.

**Comment by yubiuser** [comment]
I think it has to do with me having issues to switch to this branch. I can successfully switch, but the group management pages won't work. They try to load the lists forever. Lighttpd does not report any error.

File: scripts/pi-hole/js/utils.js [hunk 2]
```diff
@@ -248,6 +248,81 @@ function getGraphType() {
...
+      success: function (response) {
+        alProcessing.hide();
+        if (!response.success) {
+          // Failure
+          alNetworkErr.hide();
+          alCustomErr.html(response.message);
+          alFailure.fadeIn(1000);
+          setTimeout(function () {
+            alertModal.modal("hide");
+          }, 10000);
```
Comment (PromoFaux) [inline scripts/pi-hole/js/utils.js:294]: 10 seconds might be a bit long now

File: scripts/pi-hole/js/utils.js [hunk 3]
```diff
@@ -248,6 +248,81 @@ function getGraphType() {
...
+        }
+      },
+      error: function () {
+        // Network Error
+        alProcessing.hide();
+        alNetworkErr.show();
+        alFailure.fadeIn(1000);
+        setTimeout(function () {
+          alertModal.modal("hide");
+        }, 8000);
```
Comment (PromoFaux) [inline scripts/pi-hole/js/utils.js:312]: and/or 8 seconds might also be too long still.

Maybe 5?

**Review by PromoFaux — CHANGES_REQUESTED** [review]

File: scripts/pi-hole/js/utils.js (hunk 3 above)
Comment (DL6ER) [reply scripts/pi-hole/js/utils.js:312]: 5 instead of 4 doesn't really sound like much of a difference, though?

Comment (DL6ER) [reply scripts/pi-hole/js/utils.js:312]: I added a close button.

**Comment by yubiuser** [comment]
After merging dev branch, checkout was successful.
@DL6ER example as shown above is working now as expected. 

Will to more testing in the evening.

**Comment by yubiuser** [comment]
Found two issue (maybe slightly unrelated): `test.com` is blacklisted for group test only. Going to blacklist or group management/domains and trying to add `test.com` via "Add to Blacklist" results in
1) Green "Success" and "Not adding, is already on a list" (should be red and not "success") /Think I've seen a bug report somewhere
2) Reverts the group configuration back to "Default" only for `test.com` -> Expected would be to add it to default as well.

**Comment by yubiuser** [comment]
Additional: the "Add to blacklist" and "Add to whitelist" in Group Management should behave the same as the buttons in the Query log (replace instead of add). But they allow to add the same domain to back and whitelist in default group.

**Comment by DL6ER** [comment]
> they allow to add the same domain to back and whitelist in default group.

This is expected. Users will have to be able to do this in order to add an already blacklisted domain also to the whitelist (to modify its groups afterwards). So I guess that's correct *and* expected.

> Found two issue (maybe slightly unrelated)

> test.com is blacklisted for group test only. Going to blacklist or group management/domains and trying to add test.com via "Add to Blacklist" results in Green "Success" and "Not adding, is already on a list" (should be red and not "success")

This seems correct, though? It *is* already on the blacklist?

> Reverts the group configuration back to "Default" only for test.com -> Expected would be to add it to default as well.

This should not happen. It should not touch an existing entry, I'll do some research on it.

**Comment by DL6ER** [comment]
Okay, should be fixed now.

**Comment by yubiuser** [comment]
Confirm this is fixed now.

>This seems correct, though? It is already on the blacklist?

Yes. But I think maybe orange would be a better color (like warning - you forgot, you already added this)

>This is expected. Users will have to be able to do this in order to add an already blacklisted domain also to the whitelist (to modify its groups afterwards). So I guess that's correct and expected.

You're right. I was thinking if users should be "forced" to first set their special group config and added default last. But that might be too much handholding...

____

If I assign a domain to a group only and delete the group it leaves the domain in group "not selected". Is this the intended behavior? Should it be allowed to delete non-empty groups at all? Should the domain fall back to group "default"? It behaves as if it is on gravity - but trying to add it to whitelist gives the "success, already on the list" message.

____
P.S The feature of this PR is working as expected now.

**Comment by DL6ER** [comment]
> Should it be allowed to delete non-empty groups at all?

Yes, this seems correct to me.

> Should the domain fall back to group "default"?

No, this would (maybe even drastically) change the blocking behavior. The domain may have been assigned only to a single client with the special group. If it falls into "no group", then that's the same as being disabled. If it'd be moved over to "Default", however, it would possibly affect all devices on the current network.

> It behaves as if it is on gravity - but trying to add it to whitelist gives the "success, already on the list" message.

Can you expand on that? On which list is it? When a domain exists but it not assigned to any group, it should be ignored.

**Comment by yubiuser** [comment]
> Can you expand on that? On which list is it? When a domain exists but it not assigned to any group, it should be ignored.

My fault. `test.com` seems to be a bad choice of mine - it's really on one of my adlist (https://raw.githubusercontent.com/blocklistproject/Lists/master/ads.txt)

**Review by yubiuser — APPROVED** [review]

**Review by PromoFaux — APPROVED** [review]

**Pull request merged**